### PR TITLE
Iptables enhancements

### PR DESCRIPTION
**Pull request by mliudev**
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

iptables
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel 51ec35378d) last updated 2016/08/24 15:29:56 (GMT -400)
```
##### SUMMARY

This PR extends the iptables module to allow for flushing tables and chains and setting builtin table policies.

The chain parameter is no longer required if the flush parameter is specified. This is because the flush parameter will flush the entire table if a chain is not specified. Instead we check to see if either chain or flush is specified.

A minor change in the push_arguments(...) function lets us conditionally construct a rule. Not all commands require a rule, such as when the flush or policy parameters are specified. This change allows the module to ignore all other parameters if either of 'flush' or 'policy' is defined.

Output of flush command using ansible-hacking environment:

```
(env)[root@localhost ~]# test-module -m iptables.py -a "flush=true"
* including generated source, if any, saving to: /root/.ansible_module_generated
* ansiballz module detected; extracted module source to: /root/debug_dir
***********************************
RAW OUTPUT

{"chain": null, "changed": false, "rule": "", "failed": false, "state": "present", "flush": true, "invocation": {"module_args": {"comment": null, "set_counters": null, "chain": null, "out_interface": null, "limit_burst": null, "ctstate": [], "jump": null
, "to_ports": null, "flush": true, "protocol": null, "icmp_type": null, "to_destination": null, "uid_owner": null, "set_dscp_mark_class": null, "destination": null, "set_dscp_mark": null, "source": null, "state": "present", "policy": null, "match": [], "
goto": null, "fragment": null, "to_source": null, "in_interface": null, "source_port": null, "destination_port": null, "table": "filter", "reject_with": null, "limit": null, "action": "append", "ip_version": "ipv4"}}, "table": "filter", "ip_version": "ip
v4"}

***********************************
PARSED OUTPUT
{
    "chain": null,
    "changed": false,
    "failed": false,
    "flush": true,
    "invocation": {
        "module_args": {
            "action": "append",
            "chain": null,
            "comment": null,
            "ctstate": [],
            "destination": null,
            "destination_port": null,
            "flush": true,
            "fragment": null,
            "goto": null,
            "icmp_type": null,
            "in_interface": null,
            "ip_version": "ipv4",
            "jump": null,
            "limit": null,
            "limit_burst": null,
            "match": [],
            "out_interface": null,
            "policy": null,
            "protocol": null,
            "reject_with": null,
            "set_counters": null,
            "set_dscp_mark": null,
            "set_dscp_mark_class": null,
            "source": null,
            "source_port": null,
            "state": "present",
            "table": "filter",
            "to_destination": null,
            "to_ports": null,
            "to_source": null,
            "uid_owner": null
        }
    },
    "ip_version": "ipv4",
    "rule": "",
    "state": "present",
    "table": "filter"
}
```

Output from setting policy:

```
(env)[root@localhost ~]# test-module -m iptables.py -a "chain=INPUT policy=ACCEPT"
* including generated source, if any, saving to: /root/.ansible_module_generated
* ansiballz module detected; extracted module source to: /root/debug_dir
***********************************
RAW OUTPUT

{"chain": "INPUT", "changed": false, "rule": "", "failed": false, "state": "present", "flush": false, "invocation": {"module_args": {"comment": null, "set_counters": null, "chain": "INPUT", "ip_version": "ipv4", "out_interface": null, "limit_burst": null
, "ctstate": [], "jump": null, "to_ports": null, "flush": false, "protocol": null, "icmp_type": null, "to_destination": null, "uid_owner": null, "set_dscp_mark_class": null, "destination": null, "set_dscp_mark": null, "source": null, "state": "present",
"policy": "ACCEPT", "match": [], "goto": null, "fragment": null, "to_source": null, "action": "append", "destination_port": null, "table": "filter", "reject_with": null, "limit": null, "source_port": null, "in_interface": null}}, "table": "filter", "ip_v
ersion": "ipv4"}

***********************************
PARSED OUTPUT
{
    "chain": "INPUT",
    "changed": false,
    "failed": false,
    "flush": false,
    "invocation": {
        "module_args": {
            "action": "append",
            "chain": "INPUT",
            "comment": null,
            "ctstate": [],
            "destination": null,
            "destination_port": null,
            "flush": false,
            "fragment": null,
            "goto": null,
            "icmp_type": null,
            "in_interface": null,
            "ip_version": "ipv4",
            "jump": null,
            "limit": null,
            "limit_burst": null,
            "match": [],
            "out_interface": null,
            "policy": "ACCEPT",
            "protocol": null,
            "reject_with": null,
            "set_counters": null,
            "set_dscp_mark": null,
            "set_dscp_mark_class": null,
            "source": null,
            "source_port": null,
            "state": "present",
            "table": "filter",
            "to_destination": null,
            "to_ports": null,
            "to_source": null,
            "uid_owner": null
        }
    },
    "ip_version": "ipv4",
    "rule": "",
    "state": "present",
    "table": "filter"
}
```

```
```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @mliudev for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by mliudev** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @mliudev. To the current maintainers, @LinusU please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: system/iptables.py [hunk 1]
```diff
@@ -406,15 +454,21 @@ def main():
...
             icmp_type=dict(required=False, default=None, type='str'),
+            flush=dict(required=False, default=False, type='bool'),
+            policy=dict(
+                required=False,
+                default=None,
+                type='str',
+                choices=['ACCEPT', 'DROP', 'QUEUE', 'RETURN']),
         ),
         mutually_exclusive=(
             ['set_dscp_mark', 'set_dscp_mark_class'],
```
Comment (LinusU) [inline system/iptables.py:474]: Should probably `['flush', 'policy', 'chain']` here?

**Comment by LinusU** [comment]
Would like to see the `mutually_exclusive` tuple and check mode (`module.check_mode`) support to the two new modes added, otherwise it looks very good 👍 

'needs_revision'

**Comment by gregdek** [comment]
Thanks @mliudev for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by mliudev** [comment]
@LinusU That's an interesting point. 'policy' and 'flush' can both take a 'chain' argument, but should not be used together. I think probably ['flush', 'policy'] should be sufficient?

As for check mode, the iptables command doesn't support checking flush and policy options. The check command only works for rules.

**Comment by mliudev** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @mliudev. To the current maintainers, @LinusU please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by LinusU** [comment]
Makes sense 👍 

'shipit'

**Comment by gregdek** [comment]
Thanks again to @mliudev for this PR, and thanks @LinusU for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request merged**